### ISSUE TITLE
Inferred type gets narrowed to `Never` under certain conditions

### ISSUE THREAD

**Issue by shiftinv** [microsoft/pyright#4110]
**Describe the bug**
The inferred type of a variable without an explicit annotation, which gets assigned a different type later on, is being narrowed to `Never` in some conditional blocks.

**To Reproduce**
```py
class X:
    def __init__(self):
        self.value = None

    def func(self, param: int):
        reveal_type(self.value)  # > Type of "self.value" is "int | None"

        if self.value is not None:
            reveal_type(self.value)  # > Type of "self.value" is "Never"
            self.value.bit_length()

        self.value = param
```
Removing the `self.value.bit_length()` statement, or just changing it to `self.value.bit_length`, results in the type of `self.value` being narrowed to `int` correctly. This isn't specific to `bit_length` and happens with any call on `self.value`.

The same behavior can also be reproduced with functions and global variables, and as such is not specific to instance attributes.

**Expected behavior**
The type of `self.value` at the 2nd `reveal_type` should be `int`.

**VS Code extension or command-line**
cli, 1.1.277
Going back through previous versions, the first version this happens in is 1.1.243; the type gets narrowed to `int` in 1.1.242.

**Comment by erictraut** [comment]
I agree this is a bug. I suspect it's the same root cause as https://github.com/microsoft/pyright/issues/4040, but I'll need to investigate further to confirm.

**Comment by erictraut** [comment]
My hypothesis was incorrect. Those two issues are unrelated.

This issue was introduced with [this commit](https://github.com/microsoft/pyright/commit/3bce53eabf32602c8980a1911599043e19cc573c).

**Comment by shiftinv** [comment]
Yeah, I read through the other issue before and it seemed like it could possibly be related; I wasn't too certain on it being the same underlying issue, especially since this one appears to date back further. Thank you for looking into it!

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.282, which I just published. It will also be included in a future release of pylance.
